docs/memory_management.md

```diff
@@ -0,0 +1,40 @@
+## Memory management
+
+The Spark-RAPIDS plugin manages device memory to effectively allocate the limited device memory resource among concurrent tasks.
+For memory management, the plugin tracks every device memory allocation and de-allocation request during processing.
+While there is enough memory available, the allocation request succeeds and the task continues processing.
+However, when the allocation request cannot succeed due to lack of memory, the plugin pauses that thread. When all of the active tasks have at least one thread paused, the plugin starts to roll back some of those paused threads to points where all of their input data is spillable, and let the other threads try to complete. If every thread except one has been rolled back and the one remaining thread cannot still make progress, then pluging picks up one thread to split its input and try again.
+
+### State machine for OOM handler
+
+The Spark-RAPIDS plugin keeps track of the state of the individual threads. Note that one Spark task can use multiple threads during execution.
+
+The thread can have one of these states at a time:
+
+- `UNKNOWN`: the thread has not been registered with the tracking system.
+- `THREAD_RUNNING`: the thread is running normally.
+- `THREAD_ALLOC`: the thread has initiated a memory allocation.
+- `THREAD_ALLOC_FREE`: the thread has requested a memory free before the allocation completes.
+- `THREAD_BLOCKED`: the allocation is blocked due to lack of memory. The thread is waiting for enough memory to be available.
+- `THREAD_BUFN_THROW`: a deadlock has been detected as all threads are blocked, and this thread has been selected to roll back to the point where all its data is spillable.
+- `THREAD_BUFN_WAIT`: the thread has initiated the rollback.
+- `THREAD_BUFN`: the thread has rolled back and is now blocked until further notice (BUFN). The task will be unblocked once high priority tasks release enough memory.
+- `THREAD_SPLIT_THROW`: a deadlock has been detected as all threads are BUFN, and this thread has been selected to roll back, split its input, and retry.
+- `THREAD_REMOVE_THROW`: the task has been unregistered while blocked.
+
+The thread state can change based on the diagram below. Note that the thread state can transition from any state to `UNKNOWN`, but it is omitted in the diagram for brevity.
```

Comment:
Actually a thread cannot transition to UNKNOWN from THREAD_BLOCKED or THREAD_BUFN because those states indicate that the thread itself is blocked. They have to transition through THREAD_REMOVE_THROW first so that the thread is guaranteed to be woken up.

Reply:
Thanks, fixed this description.